loader/cJSON.c

```diff
@@ -951,65 +951,99 @@ cJSON *loader_cJSON_GetObjectItem(cJSON *object, const char *string) {
     return c;
 }
 
-VkResult loader_get_json(const struct loader_instance *inst, const char *filename, cJSON **json) {
+#ifdef _WIN32
+static VkResult loader_read_entire_file(const struct loader_instance *inst, const char *filename, char **out_buff) {
     FILE *file = NULL;
-    char *json_buf = NULL;
-    size_t len;
     VkResult res = VK_SUCCESS;
+    DWORD len = 0;
 
-    assert(json != NULL);
-
-    *json = NULL;
-
-#if defined(_WIN32)
     int filename_utf16_size = MultiByteToWideChar(CP_UTF8, 0, filename, -1, NULL, 0);
     if (filename_utf16_size > 0) {
         wchar_t *filename_utf16 = (wchar_t *)loader_stack_alloc(filename_utf16_size * sizeof(wchar_t));
         if (MultiByteToWideChar(CP_UTF8, 0, filename, -1, filename_utf16, filename_utf16_size) == filename_utf16_size) {
             errno_t wfopen_error = _wfopen_s(&file, filename_utf16, L"rb");
             if (0 != wfopen_error) {
                 loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0, "loader_get_json: Failed to open JSON file %s", filename);
+                res = VK_ERROR_INITIALIZATION_FAILED;
+                goto out;
             }
         }
     }
+    if (NULL == (*out_buff = (char *)loader_instance_heap_calloc(inst, len + 1, VK_SYSTEM_ALLOCATION_SCOPE_COMMAND))) {
+        loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0, "loader_get_json: Failed to read file size of JSON file %s", filename);
+        res = VK_ERROR_INITIALIZATION_FAILED;
+        goto out;
+    }
+    if (INVALID_FILE_SIZE == (len = GetFileSize(file, NULL))) {
+        loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0, "loader_get_json: Failed to allocate memory to read JSON file %s", filename);
+        res = VK_ERROR_OUT_OF_HOST_MEMORY;
+        goto out;
+    }
+    if (len != fread(*out_buff, sizeof(char), len, file)) {
+        loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0, "loader_get_json: Failed to read entire JSON file %s", filename);
+        res = VK_ERROR_INITIALIZATION_FAILED;
+        goto out;
+    }
+    (*out_buff)[len] = '\0';
+
+out:
+    if (NULL != file) {
+        fclose(file);
+    }
+    return res;
+}
 #elif COMMON_UNIX_PLATFORMS
-    file = fopen(filename, "rb");
-#else
-#warning fopen not available on this platform
-#endif
+#include <fcntl.h>
+#include <sys/stat.h>
```

Comment:
I could have sworn these headers were already included elsewhere, but even if they aren't, headers need to be done at the top of the file. It becomes rather annoying to read the code and find that the required headers are strewn about randomly in its implementation.

Reply:
Yup, good point.